crates/astria-sequencer/src/app/mod.rs

```diff
@@ -148,35 +149,171 @@ const POST_TRANSACTION_EXECUTION_RESULT_KEY: &str = "post_transaction_execution_
 type InterBlockState = Arc<StateDelta<Snapshot>>;
 
 /// This is used to identify a proposal constructed by the app instance
-/// in `prepare_proposal` during a `process_proposal` call.
-///
-/// The fields are not exhaustive, in most instances just the validator address
-/// is adequate. When running a third party signer such as horcrux however it is
-/// possible that multiple nodes are preparing proposals as the same validator
-/// address, in these instances the timestamp is used as a unique identifier for
-/// the proposal from that node. This is not a perfect solution, but it only
-/// impacts sentry nodes does not halt the network and is cheaper computationally
-/// than an exhaustive comparison.
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+/// in `prepare_proposal` during a `process_proposal`, as well as the
+/// executed block during a `finalize_block` call. We validate that
+/// all parts of the proposal match in all calls. Otherwise execution
+/// is redone.
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub(crate) struct ProposalFingerprint {
-    validator_address: account::Id,
-    timestamp: tendermint::Time,
+    sha: Option<[u8; 32]>,
+    // Hash of all data except the hash on a proposal. The hash is not available
+    // during prepare_proposal, so we need to store it separately.
+    hash: Option<Hash>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum ExecutedProposalValidationStatus {
+    Unset,
+    Valid,
+    Invalid,
 }
```

Comment:
don't like the term `validation` as we're not validating here, we're checking if the block has already been executed or not. change to `BlockExecutedStatus` or something like that?